File: backend/services/embeddingService.js
const axios = require('axios');

async function generateEmbedding(text) {
  try {
    const headers = {
      'Content-Type': 'application/json',
    };
    if (process.env.HUGGINGFACE_API_KEY) {
      headers['Authorization'] = `Bearer ${process.env.HUGGINGFACE_API_KEY}`;
    }
    
    const response = await axios.post(
      'https://router.huggingface.co/hf-inference/models/sentence-transformers/all-MiniLM-L6-v2/pipeline/feature-extraction',
      { inputs: text },
      { headers, timeout: 15000 }
    );
    
    // Hugging Face may return nested arrays: [[...embedding...]]
    let embedding = response.data;
    
    // Unwrap nested arrays until we get a flat number array
    while (Array.isArray(embedding) && Array.isArray(embedding[0])) {
      embedding = embedding[0];
    }
    
    if (Array.isArray(embedding) && typeof embedding[0] === 'number') {
      return embedding;
    }
    
    throw new Error('Invalid embedding response');
  } catch (error) {
    console.error('Embedding error:', error.message);
    throw error;
  }
}

module.exports = { generateEmbedding };

File: python_backend/app/services/embedding_service.py
import requests

from ..config import settings

HF_EMBEDDING_URL = (
    "https://router.huggingface.co/hf-inference/models/"
    "sentence-transformers/all-MiniLM-L6-v2/pipeline/feature-extraction"
)


def generate_embedding(text: str) -> list[float]:
    headers = {"Content-Type": "application/json"}
    if settings.huggingface_api_key:
        headers["Authorization"] = f"Bearer {settings.huggingface_api_key}"

    response = requests.post(
        HF_EMBEDDING_URL,
        json={"inputs": text},
        headers=headers,
        timeout=15,
    )
    response.raise_for_status()

    embedding = response.json()
    while isinstance(embedding, list) and embedding and isinstance(embedding[0], list):
        embedding = embedding[0]

    if isinstance(embedding, list) and embedding and isinstance(embedding[0], (int, float)):
        return embedding

    raise ValueError("Invalid embedding response")
